Cryptic error message when instantiating type that fails to implement interface methods

### Current Behavior

Consider the following minimal reproducer:
```
given:
buildFile << """
    abstract class Reproducer implements Named {
        @Inject
        Reproducer() {}
    }
    objects.newInstance(Reproducer)
"""

expect:
succeeds("help")
```

This fails with:
```
	at org.gradle.internal.instantiation.generator.DependencyInjectingInstantiator.doCreate(DependencyInjectingInstantiator.java:69)
	at org.gradle.internal.instantiation.generator.DependencyInjectingInstantiator.newInstance(DependencyInjectingInstantiator.java:55)
	at org.gradle.api.internal.model.DefaultObjectFactory.newInstance(DefaultObjectFactory.java:90)
	at build_1slav2hx8m13lbvv1i8qw5ki6.run(gradle/platforms/software/dependency-management/build/tmp/teŝt files/Reproducer/fs6a6/build.gradle:6)
	at org.gradle.groovy.scripts.internal.DefaultScriptRunnerFactory$ScriptRunnerImpl.run(DefaultScriptRunnerFactory.java:91)
	... 150 more
Caused by: java.lang.reflect.GenericSignatureFormatError: Signature Parse error: expected '('
	Remaining input: Ljava/lang/String;()V
	at org.gradle.internal.instantiation.generator.AbstractClassGenerator$GeneratedClassImpl$GeneratedConstructorImpl.getGenericParameterTypes(AbstractClassGenerator.java:554)
	at org.gradle.internal.instantiation.generator.DependencyInjectingInstantiator.addServicesToParameters(DependencyInjectingInstantiator.java:162)
	at org.gradle.internal.instantiation.generator.DependencyInjectingInstantiator.convertParameters(DependencyInjectingInstantiator.java:121)
	at org.gradle.internal.instantiation.generator.DependencyInjectingInstantiator.doCreate(DependencyInjectingInstantiator.java:62)
	... 154 more
```

### Expected Behavior

An error message that better guides the user to a solution

### Context (optional)

_No response_

### Steps to Reproduce

See above

### Gradle version

8.4

### Build scan URL (optional)

_No response_

### Your Environment (optional)

_No response_

--- a/platforms/core-configuration/model-core/src/main/java/org/gradle/model/internal/asm/AsmClassGeneratorUtils.java
+++ b/platforms/core-configuration/model-core/src/main/java/org/gradle/model/internal/asm/AsmClassGeneratorUtils.java
@@ -21,6 +21,7 @@ import java.lang.reflect.Constructor;
 import java.lang.reflect.GenericArrayType;
 import java.lang.reflect.Method;
 import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
 import java.lang.reflect.TypeVariable;
 import java.lang.reflect.WildcardType;
 
@@ -39,13 +40,13 @@ public class AsmClassGeneratorUtils {
      */
     public static String signature(Constructor<?> constructor, boolean addNameParameter) {
         StringBuilder builder = new StringBuilder();
-        visitFormalTypeParameters(builder, constructor.getTypeParameters());
+        visitFormalTypeParameters(constructor.getTypeParameters(), builder);
         if (addNameParameter) {
             visitType(String.class, builder);
         }
-        visitParameters(builder, constructor.getGenericParameterTypes());
+        visitParameters(constructor.getGenericParameterTypes(), builder);
         builder.append("V");
-        visitExceptions(builder, constructor.getGenericExceptionTypes());
+        visitExceptions(constructor.getGenericExceptionTypes(), builder);
         return builder.toString();
     }
 
@@ -61,29 +62,33 @@ public class AsmClassGeneratorUtils {
      */
     public static String signature(Method method) {
         StringBuilder builder = new StringBuilder();
-        visitFormalTypeParameters(builder, method.getTypeParameters());
-        visitParameters(builder, method.getGenericParameterTypes());
+        visitFormalTypeParameters(method.getTypeParameters(), builder);
+        visitParameters(method.getGenericParameterTypes(), builder);
         visitType(method.getGenericReturnType(), builder);
-        visitExceptions(builder, method.getGenericExceptionTypes());
+        visitExceptions(method.getGenericExceptionTypes(), builder);
         return builder.toString();
     }
 
-    private static void visitExceptions(StringBuilder builder, java.lang.reflect.Type[] exceptionTypes) {
+    private static void visitExceptions(Type[] exceptionTypes, StringBuilder builder) {
         for (java.lang.reflect.Type exceptionType : exceptionTypes) {
             builder.append('^');
             visitType(exceptionType, builder);
         }
     }
 
-    private static void visitParameters(StringBuilder builder, java.lang.reflect.Type[] parameterTypes) {
+    private static void visitParameters(Type[] parameterTypes, StringBuilder builder) {
         builder.append('(');
-        for (java.lang.reflect.Type paramType : parameterTypes) {
-            visitType(paramType, builder);
-        }
-        builder.append(")");
+        visitTypes(parameterTypes, builder);
+        builder.append(')');
     }
 
-    private static void visitFormalTypeParameters(StringBuilder builder, TypeVariable<?>[] typeParameters) {
+    private static void visitTypes(Type[] types, StringBuilder builder) {
+        for (Type type : types) {
+            visitType(type, builder);
+        }
+    }
+
+    private static void visitFormalTypeParameters(TypeVariable<?>[] typeParameters, StringBuilder builder) {
         if (typeParameters.length > 0) {
             builder.append('<');
             for (TypeVariable<?> typeVariable : typeParameters) {
@@ -99,58 +104,39 @@ public class AsmClassGeneratorUtils {
 
     private static void visitType(java.lang.reflect.Type type, StringBuilder builder) {
         if (type instanceof Class) {
-            Class<?> cl = (Class<?>) type;
-            if (cl.isPrimitive()) {
-                builder.append(descriptorOf(cl));
-            } else {
-                if (cl.isArray()) {
-                    builder.append(cl.getName().replace('.', '/'));
-                } else {
-                    builder.append('L');
-                    builder.append(cl.getName().replace('.', '/'));
-                    builder.append(';');
-                }
-            }
+            visitClass((Class<?>) type, builder);
         } else if (type instanceof ParameterizedType) {
-            ParameterizedType parameterizedType = (ParameterizedType) type;
-            visitNested(parameterizedType.getRawType(), builder);
-            builder.append('<');
-            for (java.lang.reflect.Type param : parameterizedType.getActualTypeArguments()) {
-                visitType(param, builder);
-            }
-            builder.append(">;");
+            visitParameterizedType((ParameterizedType) type, builder);
         } else if (type instanceof WildcardType) {
-            WildcardType wildcardType = (WildcardType) type;
-            if (wildcardType.getUpperBounds().length == 1 && wildcardType.getUpperBounds()[0].equals(Object.class)) {
-                if (wildcardType.getLowerBounds().length == 0) {
-                    builder.append('*');
-                    return;
-                }
-            } else {
-                for (java.lang.reflect.Type upperType : wildcardType.getUpperBounds()) {
-                    builder.append('+');
-                    visitType(upperType, builder);
-                }
-            }
-            for (java.lang.reflect.Type lowerType : wildcardType.getLowerBounds()) {
-                builder.append('-');
-                visitType(lowerType, builder);
-            }
+            visitWildcardType((WildcardType) type, builder);
         } else if (type instanceof TypeVariable) {
-            TypeVariable<?> typeVar = (TypeVariable<?>) type;
-            builder.append('T');
-            builder.append(typeVar.getName());
-            builder.append(';');
+            visitTypeVariable((TypeVariable<?>) type, builder);
         } else if (type instanceof GenericArrayType) {
-            GenericArrayType arrayType = (GenericArrayType) type;
-            builder.append('[');
-            visitType(arrayType.getGenericComponentType(), builder);
+            visitGenericArrayType((GenericArrayType) type, builder);
         } else {
             throw new IllegalArgumentException(String.format("Cannot generate signature for %s.", type));
         }
     }
 
-    private static void visitNested(java.lang.reflect.Type type, StringBuilder builder) {
+    private static void visitGenericArrayType(GenericArrayType type, StringBuilder builder) {
+        builder.append('[');
+        visitType(type.getGenericComponentType(), builder);
+    }
+
+    private static void visitTypeVariable(TypeVariable<?> type, StringBuilder builder) {
+        builder.append('T');
+        builder.append(type.getName());
+        builder.append(';');
+    }
+
+    private static void visitParameterizedType(ParameterizedType type, StringBuilder builder) {
+        visitRawType(type.getRawType(), builder);
+        builder.append('<');
+        visitTypes(type.getActualTypeArguments(), builder);
+        builder.append(">;");
+    }
+
+    private static void visitRawType(java.lang.reflect.Type type, StringBuilder builder) {
         if (type instanceof Class) {
             Class<?> cl = (Class<?>) type;
             if (cl.isPrimitive()) {
@@ -161,6 +147,41 @@ public class AsmClassGeneratorUtils {
             }
         } else {
             visitType(type, builder);
+        }
+    }
+
+    private static void visitWildcardType(WildcardType type, StringBuilder builder) {
+        Type[] upperBounds = type.getUpperBounds();
+        if (upperBounds.length == 1 && upperBounds[0].equals(Object.class)) {
+            if (type.getLowerBounds().length == 0) {
+                builder.append('*');
+                return;
+            }
+        } else {
+            visitTypesWithPrefix('+', upperBounds, builder);
+        }
+        visitTypesWithPrefix('-', type.getLowerBounds(), builder);
+    }
+
+    private static void visitTypesWithPrefix(char prefix, Type[] types, StringBuilder builder) {
+        for (Type upperType : types) {
+            builder.append(prefix);
+            visitType(upperType, builder);
+        }
+    }
+
+    private static void visitClass(Class<?> type, StringBuilder builder) {
+        if (type.isPrimitive()) {
+            builder.append(descriptorOf(type));
+        } else {
+            String binaryName = type.getName().replace('.', '/');
+            if (type.isArray()) {
+                builder.append(binaryName);
+            } else {
+                builder.append('L');
+                builder.append(binaryName);
+                builder.append(';');
+            }
         }
     }
 

--- a/platforms/core-configuration/model-core/src/main/java/org/gradle/model/internal/asm/AsmClassGeneratorUtils.java
+++ b/platforms/core-configuration/model-core/src/main/java/org/gradle/model/internal/asm/AsmClassGeneratorUtils.java
@@ -41,13 +41,19 @@ public class AsmClassGeneratorUtils {
     public static String signature(Constructor<?> constructor, boolean addNameParameter) {
         StringBuilder builder = new StringBuilder();
         visitFormalTypeParameters(constructor.getTypeParameters(), builder);
-        if (addNameParameter) {
-            visitType(String.class, builder);
-        }
-        visitParameters(constructor.getGenericParameterTypes(), builder);
-        builder.append("V");
+        visitConstructorParameters(constructor.getGenericParameterTypes(), addNameParameter, builder);
+        builder.append('V');
         visitExceptions(constructor.getGenericExceptionTypes(), builder);
         return builder.toString();
+    }
+
+    private static void visitConstructorParameters(Type[] parameterTypes, boolean addNameParameter, StringBuilder builder) {
+        builder.append('(');
+        if (addNameParameter) {
+            visitClass(String.class, builder);
+        }
+        visitTypes(parameterTypes, builder);
+        builder.append(')');
     }
 
     public static String getterSignature(java.lang.reflect.Type returnType) {

--- a/platforms/core-configuration/model-core/src/main/java/org/gradle/model/internal/asm/AsmClassGeneratorUtils.java
+++ b/platforms/core-configuration/model-core/src/main/java/org/gradle/model/internal/asm/AsmClassGeneratorUtils.java
@@ -28,12 +28,6 @@ import java.lang.reflect.WildcardType;
 import static org.objectweb.asm.Type.getType;
 
 public class AsmClassGeneratorUtils {
-    /**
-     * Generates the signature for the given constructor
-     */
-    public static String signature(Constructor<?> constructor) {
-        return signature(constructor, false);
-    }
 
     /**
      * Generates the signature for the given constructor, optionally adding a `name` parameter before all other parameters.

--- a/subprojects/core/src/integTest/groovy/org/gradle/api/internal/model/ObjectFactoryIntegrationTest.groovy
+++ b/subprojects/core/src/integTest/groovy/org/gradle/api/internal/model/ObjectFactoryIntegrationTest.groovy
@@ -796,7 +796,7 @@ class ObjectFactoryIntegrationTest extends AbstractIntegrationSpec {
     def "plugin can create ExtensiblePolymorphicDomainObjectContainer instances using named managed types"() {
         given:
         buildFile """
-            interface BaseThing extends Named { 
+            interface BaseThing extends Named {
                 Property<Integer> getValue()
             }
             interface ThingA extends BaseThing { }
@@ -805,10 +805,10 @@ class ObjectFactoryIntegrationTest extends AbstractIntegrationSpec {
             def container = project.objects.polymorphicDomainObjectContainer(BaseThing)
             container.registerBinding(ThingA, ThingA)
             container.registerBinding(ThingB, ThingB)
-            container.register("a", ThingA) { 
+            container.register("a", ThingA) {
                 value = 0
             }
-            container.register("b", ThingB) { 
+            container.register("b", ThingB) {
                 value = 1
             }
             assert container.size() == 2

--- a/subprojects/core/src/integTest/groovy/org/gradle/api/internal/model/ObjectFactoryIntegrationTest.groovy
+++ b/subprojects/core/src/integTest/groovy/org/gradle/api/internal/model/ObjectFactoryIntegrationTest.groovy
@@ -16,6 +16,7 @@
 package org.gradle.api.internal.model
 
 import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import spock.lang.Issue
 
 class ObjectFactoryIntegrationTest extends AbstractIntegrationSpec {
     def "plugin can create instances of class using injected factory"() {
@@ -195,6 +196,24 @@ class ObjectFactoryIntegrationTest extends AbstractIntegrationSpec {
 
         expect:
         succeeds()
+    }
+
+    @Issue('https://github.com/gradle/gradle/issues/27108')
+    def "plugin can try to create instance of abstract class that implements Named and get nice error message for missing name argument"() {
+        given:
+        buildFile """
+            abstract class Thing implements Named {
+                @Inject Thing() {}
+            }
+            objects.newInstance(Thing)
+        """
+
+        expect:
+        fails()
+
+        and:
+        failure.assertHasCause("Could not create an instance of type Thing.")
+        failure.assertHasCause("Unable to determine constructor argument #1: missing parameter of type String, or no service of type String.")
     }
 
     def "fails when abstract method cannot be implemented"() {

